### PR TITLE
Add trailing nl for `display(::TextDisplay, ...)`

### LINKED ISSUE #43766: Display linebreaks, `TextDisplay` vs `REPLDisplay`
As explained on [discourse](https://discourse.julialang.org/t/display-and-line-breaks-in-textdisplay-and-lineeditrepl/73459), we (ping @mkitti) feel that there might be an inconsistency in `display(...)` between [`TextDisplay`](https://github.com/JuliaLang/julia/blob/f8f42ec0370b72072498c1b79eb5cf4e6c6864e6/base/multimedia.jl#L242) and [`REPLDisplay`](https://github.com/JuliaLang/julia/blob/f8f42ec0370b72072498c1b79eb5cf4e6c6864e6/stdlib/REPL/src/REPL.jl#L268).

Does anyone know why there is no trailing newline for `TextDisplay` ?

Would the following patch be acceptable ?

```diff
--- a/base/multimedia.jl	2022-01-11 21:00:42.765702329 +0100
+++ b/base/multimedia.jl	2022-01-11 21:02:53.377249363 +0100
@@ -239,14 +239,14 @@
 struct TextDisplay <: AbstractDisplay
     io::IO
 end
-display(d::TextDisplay, M::MIME"text/plain", @nospecialize x) = show(d.io, M, x)
+display(d::TextDisplay, M::MIME"text/plain", @nospecialize x) = (show(d.io, M, x); println(d.io))
 display(d::TextDisplay, @nospecialize x) = display(d, MIME"text/plain"(), x)
 
 # if you explicitly call display("text/foo", x), it should work on a TextDisplay:
 displayable(d::TextDisplay, M::MIME) = istextmime(M)
 function display(d::TextDisplay, M::MIME, @nospecialize x)
     displayable(d, M) || throw(MethodError(display, (d, M, x)))
-    show(d.io, M, x)
+    show(d.io, M, x); println(d.io)
 end
 
 import Base: close, flush
```

### PR DESCRIPTION
**Pull request by t-bltg**
Fix https://github.com/JuliaLang/julia/issues/43766.

Test pass locally on `1.6.5` and `1.7.1` builds on `linux`, let's check `CI` now ...

It'd be great (if possible) to have this land in `1.6` and `1.7`.

### CODE REVIEW

File: base/client.jl [hunk 1]
```diff
@@ -142,7 +142,6 @@ function eval_user_input(errio, @nospecialize(ast), show_value::Bool)
                         @error "Evaluation succeeded, but an error occurred while displaying the value" typeof(value)
                         rethrow()
                     end
-                    println()
```
Comment (t-bltg) [inline base/client.jl:145]: This removal compensates the newline introduced in `invokelatest(display, value)`.

**Comment by mkitti** [comment]
```
Test Failed at /usr/home/julia/buildbot/w2_tester/tester_freebsd64/build/share/julia/stdlib/v1.8/DelimitedFiles/test/runtests.jl:327
  Expression: String(take!(d.io)) == "3,1,4\n"
   Evaluated: "3,1,4\n\n" == "3,1,4\n"
```

**Comment by t-bltg** [comment]
Thanks, I forgot to test the `stdlib` modules (only `Base`), will fix that asap.

**Review by JeffBezanson — APPROVED** [review]

File: stdlib/DelimitedFiles/test/runtests.jl [hunk 2]
```diff
@@ -321,10 +321,10 @@ end
 # issue #11484: useful error message for invalid readdlm filepath arguments
 @test_throws ArgumentError readdlm(tempdir())
 
-# displaying as text/csv
-let d = TextDisplay(IOBuffer())
-    display(d, "text/csv", [3 1 4])
-    @test String(take!(d.io)) == "3,1,4\n"
+# showing as text/csv
+let d = TextDisplay(PipeBuffer())
+    show(d.io, "text/csv", [3 1 4])
```
Comment (t-bltg) [inline stdlib/DelimitedFiles/test/runtests.jl:326]: I think changing this test to using `show(...)` instead of `display` is the best move here.

**Comment by xitology** [comment]
The change itself may be reasonable, but it's going to break *all* the code that uses `display` for output.

**Comment by mkitti** [comment]
> The change itself may be reasonable, but it's going to break _all_ the code that uses `display` for output.

I would be interested in learning more about code that uses `display` for output using `TextDisplay`. Do you have examples?

**Comment by xitology** [comment]
I use `display` all the time in command-line scripts to print multi-line data. I have to add an extra `println()` after each `display()` call, which I admit is irritating.

Here is my package where this change broke the test suite:
https://github.com/MechanicalRabbit/NarrativeTest.jl/runs/4831849293?check_suite_focus=true
The specific line in the package that calls `display`:
https://github.com/MechanicalRabbit/NarrativeTest.jl/blob/master/src/NarrativeTest.jl#L380

**Comment by t-bltg** [comment]
> I use display all the time in command-line scripts to print multi-line data. I have to add an extra println() after each display() call, which I admit is irritating.

I guess you should use `show` instead of `display`.

**Comment by xitology** [comment]
> > I use display all the time in command-line scripts to print multi-line data. I have to add an extra println() after each display() call, which I admit is irritating.
> 
> I guess you should use `show` instead of `display`.

`show(stdout, "text/plain", x)` is a bit mouthful compared with `display(x)`.

**Comment by t-bltg** [comment]
IMO having a consistent behavior for `display` using either `TextDisplay` and `REPLDisplay` (as was the motivation of this PR) is the right move.

Isn't `show(stdout, x)` sufficient in your case ?

**Comment by xitology** [comment]
> IMO having a consistent behavior for `display` using either `TextDisplay` or `REPLDisplay` (as was the motivation of this PR) is the right move.
> 
> Isn't `show(stdout, x)` sufficient in your case ?

It gives a different output:
```julia
julia> show(stdout, [1,2,3])
[1, 2, 3]
julia> show(stdout, "text/plain", [1,2,3])
3-element Vector{Int64}:
 1
 2
 3
julia> display([1,2,3])
3-element Vector{Int64}:
 1
 2
 3

julia> show(stdout, Dict("key" => "value"))
Dict("key" => "value")
julia> show(stdout, "text/plain", Dict("key" => "value"))
Dict{String, String} with 1 entry:
  "key" => "value"
julia> display(Dict("key" => "value"))
Dict{String, String} with 1 entry:
  "key" => "value"

```

**Comment by t-bltg** [comment]
I agree that it is painful to fix these things. But defining `disp(x) = show(stdout, "text/plain", x)` is always possible if you want to avoid using `show(stdout, "text/plain", x)`.

**Comment by xitology** [comment]
My point is that it is not a bugfix but a breaking change and should be treated as such.

**Comment by t-bltg** [comment]
> My point is that it is not a bugfix but a breaking change and should be treated as such.

I'd say it's debatable whether it is a bug fix or a breaking change.

**Comment by mkitti** [comment]
Perhaps we should add something in [Compat.jl](https://github.com/JuliaLang/Compat.jl)?

**Pull request merged**